website/docs/API.md

```diff
@@ -515,22 +517,96 @@ Defined as:
 ```jsx
 function waitFor<T>(
   expectation: () => T,
-  { timeout: number = 4500, interval: number = 50 }
+  { timeout: number = 1000, interval: number = 50 }
 ): Promise<T> {}
 ```
 
-Waits for non-deterministic periods of time until your element appears or times out. `waitFor` periodically calls `expectation` every `interval` milliseconds to determine whether the element appeared or not.
+Should be used when testing asynchronous code or code with timers. `waitFor` allows to wait for some expectation (e.g. the presence of an element on the screen). The `expectation` param is a callback that should throw until what you're waiting for has realized.
+
+For instance, this is how we'd wait for an element to appear:
+
+```tsx
+await waitFor(() => expect(screen.getByText('Banana ready').toBeTruthy()))
+```
+
+What `waitFor` does is executing `expectation()` at intervals of duration equal to the optional `interval` param (default is 50ms) until it doesn't throw anymore. It will throw if it reaches its timeout, defined by the optional `timeout` param (default is 1000ms). If `expectation()` does not throw then `waitFor` will resolve right away.
+
+```tsx
+// won't throw
+await waitFor(() => false);
+```
+
+`waitFor` only rejects when it reaches its timeout so if you don't await it it will never throw.
 
 ```jsx
-import { render, screen, waitFor } from '@testing-library/react-native';
+// won't throw
+waitFor(() => expect(1).toBe(2))
+```
 
-test('waiting for an Banana to be ready', async () => {
-  render(<Banana />);
+:::note
+You can enforce awaiting waitFor by using the [await-async-utils](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/await-async-utils.md) rule from [eslint-plugin-testing-library](https://github.com/testing-library/eslint-plugin-testing-library).
+:::
 
-  await waitFor(() => screen.getByText('Banana ready'));
-});
+Since `waitFor` is likely to run `expectation()` multiple times, it is highly recommended not to have any side effects in `waitFor`.
+
+```jsx
+await waitFor(() => {
+  // button will be pressed on each waitFor iteration
+  fireEvent.press(screen.getByText('press me'))
+  expect(screen.getByText('pressed')).toBeTruthy()
+})
 ```
 
+:::note
+Not using fireEvent in `waitFor` be enforced with the rule [no-wait-for-side-effects](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/no-wait-for-side-effects.md).
+:::
+
+### waitFor with jest fake timers
```

Comment:
I love the detailed explanation but I think a big part of this new doc should go in the Advanced Section like [Understanding the act function](https://callstack.github.io/react-native-testing-library/docs/understanding-act). When you look at the rest of the docs, it's very concise and does not go too much in details about the implementation, wdyt?

Reply:
I agree this is not at the right place but I'm not sure either that moving this doc to the advanced section is the best solution. What I don't like about advanced section is that users are way less likely to read it as it seems unnecessary to correctly use the library. I'm not exactly sure what should go in advanced section or not but I think understanding how waitFor works is both too valuable and not too complex to be in advanced section. Also the detailed explanations can easily be skipped so I don't think it's an issue for this doc to be in a single place.

Maybe the API page is meant to only give available APIs and a brief explanation but if that's the case if think more sections should be added to cover stuff like this more in detail. For instance this could be in a section about testing asynchronous code or code with timers. I feel like right now there is a huge list of available APIs but it's kind of hard to know what these APIs allow to do. For instance if I don't know how to test asynchronous code, I will need to read through the API page and find the waitFor section, and even then it may not be detailed enough because this page is quite succinct. 

What I'm thinking of right now would be additional sections like 'render component', 'use queries', 'trigger events', 'test asynchronous code' for instance, that would be under the guide section and each of these sections would provide more detailed examples of the API usage.

@MattAgn  @mdjastrzebski  @AugustinLF wdyt ?

Reply:
I like your proposition to have more `guides` section, could be useful to many people

Reply:
I think that's a problem we have with our whole doc, if you look at [RTL's docs](https://testing-library.com/docs/dom-testing-library/api-async), they have both API docs and something more detailed.

With our current structure, I'd suggest to keep this in `API.md`, advanced section is too hidden.